contracts/interchain-token/src/contract.rs

```diff
@@ -20,27 +46,171 @@ impl InterchainToken {
+                return Err(ContractError::NotAuthorizedMinter);
+            }
+        }
+
+        extend_instance_ttl(&env);
+
+        receive_balance(&env, to.clone(), amount);
+        TokenUtils::new(&env).events().mint(admin, to, amount);
 
         Ok(())
     }
+
+    pub fn set_admin(env: Env, new_admin: Address) {
+        let admin = admin(&env);
+        admin.require_auth();
+
+        extend_instance_ttl(&env);
+
+        env.storage().instance().set(&DataKey::Admin, &admin);
+        TokenUtils::new(&env)
+            .events()
+            .set_admin(admin.clone(), new_admin.clone());
+
+        event::set_admin(&env, admin, new_admin);
+    }
+
+    pub fn add_minter(env: &Env, minter: Address) {
+        admin(env).require_auth();
+
+        env.storage()
+            .persistent()
+            .set(&DataKey::Minter(minter.clone()), &true);
+
+        event::add_minter(env, minter);
+    }
+}
+
+#[contractimpl]
+impl token::Interface for InterchainToken {
+    fn allowance(env: Env, from: Address, spender: Address) -> i128 {
+        extend_instance_ttl(&env);
+        read_allowance(&env, from, spender).amount
+    }
+
+    fn approve(env: Env, from: Address, spender: Address, amount: i128, expiration_ledger: u32) {
+        from.require_auth();
+
+        check_nonnegative_amount(amount);
+        extend_instance_ttl(&env);
+
+        write_allowance(
+            &env,
+            from.clone(),
+            spender.clone(),
+            amount,
+            expiration_ledger,
+        );
+
+        TokenUtils::new(&env)
+            .events()
+            .approve(from, spender, amount, expiration_ledger);
+    }
+
+    fn balance(env: Env, id: Address) -> i128 {
+        extend_instance_ttl(&env);
+        read_balance(&env, id)
+    }
+
+    fn transfer(env: Env, from: Address, to: Address, amount: i128) {
+        from.require_auth();
+
+        check_nonnegative_amount(amount);
+        extend_instance_ttl(&env);
+
+        spend_balance(&env, from.clone(), amount);
+        receive_balance(&env, to.clone(), amount);
+        TokenUtils::new(&env).events().transfer(from, to, amount);
+    }
+
+    fn transfer_from(env: Env, spender: Address, from: Address, to: Address, amount: i128) {
+        spender.require_auth();
+
+        check_nonnegative_amount(amount);
+        extend_instance_ttl(&env);
+
+        spend_allowance(&env, from.clone(), spender, amount);
+        spend_balance(&env, from.clone(), amount);
+        receive_balance(&env, to.clone(), amount);
+        TokenUtils::new(&env).events().transfer(from, to, amount)
+    }
+
+    fn burn(env: Env, from: Address, amount: i128) {
+        from.require_auth();
+
+        check_nonnegative_amount(amount);
+        extend_instance_ttl(&env);
+
+        spend_balance(&env, from.clone(), amount);
+        TokenUtils::new(&env).events().burn(from, amount);
+    }
+
+    fn burn_from(env: Env, spender: Address, from: Address, amount: i128) {
```

Comment:
in the future, i recommend splitting the PR more (e.g. basic token functionality vs ITS specific mint/admin/setup stuff) so it's faster to review/merge